compiler-core/src/type_/expression.rs

```diff
@@ -391,13 +391,20 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
             self.environment.warnings.emit(Warning::UnusedLiteral {
                 location: discarded.location(),
             });
-        }
-        if discarded.type_().is_result() {
+        } else if discarded.type_().is_result() {
             self.environment
                 .warnings
                 .emit(Warning::ImplicitlyDiscardedResult {
                     location: discarded.location(),
                 });
+        } else if discarded.is_binop()
+            || discarded.is_record_constructor()
+            || discarded.is_record_access()
+            || discarded.is_record_update()
```

Comment:
What do you think of something like `.is_pure_value_constructor()` or something like that? The if new variants are added the compiler will prompt us to decide if it's pure, while without pattern matching it won't.

Reply:
I think that's a great idea! I'll add it